Doc/library/token.rst

```diff
@@ -44,40 +44,151 @@ functions.  The functions mirror definitions in the Python C header files.
+   of Python code.
 
 .. data:: NL
-   :noindex:
 
-   Token value used to indicate a non-terminating newline.  The
-   :data:`NEWLINE` token indicates the end of a logical line of Python code;
-   ``NL`` tokens are generated when a logical line of code is continued over
-   multiple physical lines.
+   Token value used to indicate a non-terminating newline.
+   :data:`!NL` tokens are generated when a logical line of code is continued
+   over multiple physical lines. The parser ignores :data:`!NL` tokens.
+
+.. data:: INDENT
+
+   Token value used at the beginning of a :ref:`logical line <logical-lines>`
+   to indicate the start of an :ref:`indented block <indentation>`.
+
+.. data:: DEDENT
+
+   Token value used at the beginning of a :ref:`logical line <logical-lines>`
+   to indicate the end of an :ref:`indented block <indentation>`.
+
+.. data:: FSTRING_START
+
+   Token value used to indicate the beginning of a
+   :ref:`f-string <f-strings>`.
+
+   .. impl-detail::
+
+      The token string includes the prefix and the opening quote, but none
+      of the contents of the literal.
+
+.. data:: FSTRING_MIDDLE
+
+   Token value used for literal text inside an :ref:`f-string <f-strings>`,
+   including format specifications.
+
+   .. impl-detail::
+
+      Replacement fields (that is, the non-literal parts of f-strings) use
+      the same tokens as other expressions, and are delimited by
+      :data:`LBRACE`, :data:`RBRACE`, :data:`EXCLAMATION` and :data:`COLON`
+      tokens.
+
+.. data:: FSTRING_END
+
+   Token value used to indicate the end of a :ref:`f-string <f-strings>`.
+
+   .. impl-detail::
 
+      The token string contains the closing quote.
+
+.. data:: ENDMARKER
+
+   Token value that indicates the end of input.
+   Used in :ref:`top-level grammar rules <top-level>`.
 
 .. data:: ENCODING
 
    Token value that indicates the encoding used to decode the source bytes
    into text. The first token returned by :func:`tokenize.tokenize` will
    always be an ``ENCODING`` token.
 
+   .. impl-detail::
+
+      This token type isn't used by the C tokenizer but is needed for
+      the :mod:`tokenize` module.
+
+
+The following token types are not produced by the :mod:`tokenize` module,
+and are defined for special uses in the tokenizer or parser:
+
+.. data:: TYPE_IGNORE
+
+   Token value indicating that a ``type: ignore`` comment was recognized.
+   Such tokens are produced instead of regular :data:`COMMENT` tokens only when
+   :func:`ast.parse` is invoked with ``type_comments=True``.
 
 .. data:: TYPE_COMMENT
-   :noindex:
 
-   Token value indicating that a type comment was recognized.  Such
-   tokens are only produced when :func:`ast.parse` is invoked with
-   ``type_comments=True``.
+   Token value indicating that a type comment was recognized.
+   Such tokens are produced instead of regular :data:`COMMENT` tokens only when
+   :func:`ast.parse` is invoked with ``type_comments=True``.
```

Comment:
```suggestion
   Token value indicating that a type comment was recognized.
   Such tokens are produced instead of regular :data:`COMMENT` tokens only
   with the :data:`~ast.PyCF_TYPE_COMMENTS` flag.
```